text/0000-translateable-ui.md

```diff
@@ -0,0 +1,128 @@
+        height: 50.,
+        transparent: true,
+    ),
+    text: (
+        text: "loading-text",
+        translate: true,
+        font_size: 25.,
+        color: (1., 1., 1., 1.),
+        font: File("font/square.ttf", Ttf, ()),
+    ),
+)
+```
+
+```ftl
+loading-text = Loading, please wait!
+```
+
+```rust
+world.exec(|mut creator: UiCreator| {
+    creator
+        .build("prefabs/my_ui.ron")
+        .with_locales("resources/en_US.ftl")
+        .create();
+});
+
+// or, if no prefabs are used
+
+UiTranslator.translate(world, locale_storage);
+```
+
+
+# Reference-Level Explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The original text is the key, and the new text is the translation from the resource file. The translation is done by checking a marker field or component (`translate` field in above example) and match the key against a storage. The translation can either be done by a convenience method on the UiCreator directly, or by calling a translation function directly, which checks the world for translate-able components once.
+
+If no value is found to a key, the text should not be replaced. This way, at least the key is visible and a developer (or QA) knows which translation is still missing.
+
+In order to make the binding between prefab and resource easy, the creator can be extended with a builder pattern, which can add a path or a handle (for simple re-use).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I do not see any drawbacks for the proposal itself, since translation is expected of modern games and this RFC adds a simple way to implement it from a game-dev's perspective.
+
+This RFC does not take non-ui texts into account, though, which might also need translation. The above approach is hard to generalize for any component with a text field, and I cannot come up with a good solution to the problem - which might be due to my limited knowledge of Rust and Amethyst.
+
+# Rationale and Alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As a web-dev working on web apps, I usually cache the locale file and fill the HTML with attributes marking elements for translation, or directly call a function which returns a translation to a given key. It's quite a simple way and state-of-the art as far as I know (see Angular).
+
+Instead of going with the above described apporach for a technical implementation, the translation could be implemented by caching all entities with a `UiText` component on UI creation, and as soon as the locale resource is loaded, replace the text.
+
+# Prior Art
+[prior-art]: #prior-art
+
+Similarities to how locales are inserted in web frameworks.
+
+```html
+<h3 i18n>loading-text</h3>
+```
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Is this the easiest way to combine prefabs and localization?
+- How can it be generalized for any (custom?) component with text? Can we use components as markers and traits for that?
```

Comment:
pub struct Localized {
  key: String,
  applied: bool,
}

And a system doing the translation from Localized to UiText + a system for your custom components